design/sequence/scheduler-low.puml

```diff
@@ -0,0 +1,86 @@
+@startuml
+
+box "Central Server" #Lightblue
+    participant ":Database"
+    participant ":Scheduler"
+    participant "job:Job"
+    participant ":Dispatcher"
+end box
+
+
+":Database" -> ":Scheduler" ++: reschedule()
+":Scheduler" -> ":Database" ++: get_changed_job()
+":Database" --> ":Scheduler" --: job: Job
```

Comment:
But the Scheduler needs to know if a job has been killed or finished in order to react accordingly. There's no need to reschedule all Jobs just because one job has been killed or finished, we just start the first job in the queue that can run on the newly available work machine.

Reply:
We can have multiple jobs per machine. So it can happen that the killed/done job is very big, and after it is removed we have enough place to run multiple new tasks. Also, we need to consider paused jobs which have not changed at all.

Reply:
Just to clarify, I believe that giving the scheduler the complete list of jobs which are not done or killed is much better, because then the scheduler decides what happens and what won't happen.

Reply:
>But the Scheduler needs to know if a job has been killed or finished in order to react accordingly.

This should be expressed by the state of the Job.

>There's no need to reschedule all Jobs just because one job has been killed or finished, we just start the first job in the queue that can run on the newly available work machine.

Incorrect.
When a Job finishes any number of other Jobs can take its place.
Consider the following example:
* A Job with 4 threads finishes.
* There are 4 waiting Jobs that need 1 thread each.
In the above case all four waiting jobs should be scheduled even though only one Job finishes.
We need to consider that Jobs use varying amounts of resources and can therefore **not** be replaced in a 1:1 manner.
There are also more complicated cases where you for example have several paused Jobs that could be resumed and need to pick only some of them to resume.
This makes writing a method for modifying a schedule complicated.

However, we already need to write a method for creating a schedule in the first place.
Since we do not have to write efficient code the easiest way to modify a schedule will be to modify the Database and to then completely rebuild the schedule.
This approach will save us roughly 50% of the work during the implementation phase.